### PR TITLE
Rework the TF DALIDataset input API

### PR DESCRIPTION
**Pull request by klecki**
Change from separate tuples for every
parameter to a dictionary that maps
the input names to the actual inputs.

Adjust docs and tests.

TODO:
- [x] Automatic layout handling from Pipeline.

Signed-off-by: Krzysztof Lecki <klecki@nvidia.com>

#### Why we need this PR?
Improve API of TF DALI Dataset.

#### What happened in this PR? 
 - What solution was applied:
     
Change from separate tuples for every
parameter to a dictionary that maps
the input names to the actual inputs.

Adjust docs and tests.

 - Affected modules and functionalities:
     tf plugin
 - Key points relevant for the review:
     
 - Validation and testing:
     CI
 - Documentation (including examples):
     Yes

**JIRA TASK**: *[Use DALI-2144 or NA]*

### CODE REVIEW

**Comment by klecki** [comment]
!build

File: dali/python/nvidia/dali/plugin/tf.py [hunk 1]
```diff
@@ -59,26 +62,31 @@
 
 Parameters
 ----------
-    input_datasets : tf.data.Dataset or tuple of tf.data.Dataset, optional, default = None
-        input datasets to the DALI Pipeline. When provided, user must specify the ``input_names``
-        as well to provide the mapping of input datasets to ``External Source`` nodes.
+    input_datasets : dict[str, tf.data.Dataset] or dict[str, nvidia.dali.plugin.tf.experimental.input]
+        input datasets to the DALI Pipeline. It must be provided as a dictionary mapping from
+        the names of the ``External Source`` nodes to the datasets objects intended as inputs
```
Comment (jantonguirao) [inline dali/python/nvidia/dali/plugin/tf.py:67]: nitpick: "intended as inputs for those nodes" seems a bit redundant

File: dali/python/nvidia/dali/plugin/tf.py [hunk 2]
```diff
@@ -426,15 +510,37 @@ def __init__(self, pipeline, **kwargs):
...
 
-
-        current_module = sys.modules[__name__]
-        experimental_module = _internal.get_submodule(current_module, "experimental")
-        DALIDatasetWithInputs.__module__ = experimental_module
         DALIDatasetWithInputs.__doc__ = _experimental_dataset_docstring
-        setattr(experimental_module, "DALIDatasetWithInputs", DALIDatasetWithInputs)
+        _insert_experimental_member(DALIDatasetWithInputs, "DALIDatasetWithInputs")
+
+        class input:
```
Comment (jantonguirao) [inline dali/python/nvidia/dali/plugin/tf.py:516]: why not `Input`? I personally got confused when I saw `experimental.input` earlier, as I thought it was an instance

**Comment by klecki** [comment]
I'm thinking whether to keep the dictionary with name -> input mapping, or to force the use of `fn.external_source(source=tf.data.Dataset)`

**Review by awolant — APPROVED** [review]

File: dali/python/nvidia/dali/plugin/tf.py (hunk 1 above)
Comment (klecki) [reply dali/python/nvidia/dali/plugin/tf.py:67]: Removed.

File: dali/python/nvidia/dali/plugin/tf.py [hunk 3]
```diff
@@ -281,60 +293,132 @@ def __init__(self,
...
+                if not is_dataset_only and not isinstance(input_value, experimental.input):
+                    raise TypeError(error_str + (". Expected the values of the dictionary " +
+                                                 "(representing the inputs) " +
+                                                 " to be of type `tf.data.Dataset` or " +
+                                                 "`nvidia.dali.plugin.tf.input` got: `{}` of " +
+                                                 "type: {} instead.").format(input_value, type(input_value)))
+
+                # there is External Source with name equal to `input_name`
+                if input_name not in name_es_map.keys():
+                    raise ValueError(("Did not find an External Source node with nam=' {}' in " +
```
[the remark this answers is not in the document]
Comment (klecki) [reply dali/python/nvidia/dali/plugin/tf.py:356]: Done

File: dali/python/nvidia/dali/plugin/tf.py [hunk 4]
```diff
@@ -281,60 +293,132 @@ def __init__(self,
...
+                                                                      list(name_es_map.keys())))
+
+                input_names_list.append(input_name)
+                input_datasets_list.append(_get_dataset(input_value))
+
+                # TODO(klecki): Do we want all Python-only ES parameters to be overridable here?
+                if is_dataset_only or input_value.layout is None:
+                    input_layouts_list.append(self._get_layout_from_pipeline(input_name, name_es_map))
+                else:
+                    print("HERE: ", input_value, input_value.layout)
```
[the remark this answers is not in the document]
Comment (klecki) [reply dali/python/nvidia/dali/plugin/tf.py:369]: That's a rather good guess :)

File: dali/python/nvidia/dali/plugin/tf.py [hunk 5]
```diff
@@ -281,60 +293,132 @@ def __init__(self,
...
+            if len(non_matched) != 0:
+                raise ValueError(("Found External Source nodes in the Pipeline, that were not " +
+                                  "assigned any inputs. Nodes without inputs: \n{}.\nNodes that " +
+                                  "were assigned inputs:\n{}.").format(list(non_matched), list(input_datasets.keys())))
+
+            self._input_datasets = tuple(input_datasets_list)
+            self._input_names = tuple(input_names_list)
+            self._input_layouts = tuple(input_layouts_list)
+
+        def _assert_pipeline_instace(self):
```
[the remark this answers is not in the document]
Comment (klecki) [reply dali/python/nvidia/dali/plugin/tf.py:383]: Done

File: dali/python/nvidia/dali/plugin/tf.py (hunk 2 above)
Comment (klecki) [reply dali/python/nvidia/dali/plugin/tf.py:516]: Done

**Review by jantonguirao — APPROVED** [review]

**Comment by klecki** [comment]
!build

**Comment by klecki** [comment]
!build

**Pull request merged**